R/MsExperiment-db.R

```diff
@@ -0,0 +1,166 @@
+#' database.
+#'
+#' This requires that the MS data of the experiment is *represented* by a
+#' `MsBackendSql` backend (see help on the `createMsBackendSqlDatabase` or the
+#' *MsBackendSql* package vignette for more information on how to create or
+#' use such SQL databases).
+#'
+#' @param x `MsExperiment` from which sample annotations should be written
+#'     to the database.
+#'
+#' @author Johannes Rainer, Laurent Gatto
+#'
+#' @importFrom DBI dbExistsTable dbRemoveTable dbWriteTable dbDisconnect
+#'
+#' @export
+#'
+#' @examples
+#'
+#' library(MsExperiment)
+#'
+#' ## Create a MsBackendSql database from two mzML files.
+#' ## Connect first to an empty SQLite database (for the example we create
+#' ## a database in a temporary file).
+#' library(RSQLite)
+#' sqlite_db <- tempfile()
+#' con <- dbConnect(SQLite(), sqlite_db)
+#'
+#' ## Define the files from which we import the data
+#' fls <- dir(system.file("sciex", package = "msdata"), pattern = "mzML",
+#'     full.names = TRUE)
+#'
+#' ## Create a MsBackendSql database containing the full MS data
+#' library(MsBackendSql)
+#' createMsBackendSqlDatabase(con, fls)
+#'
+#' ## Note: alternatively it would be possible to first import the MS data
+#' ## to a `Spectra` object and then change the backend to a `MsBackendSql`
+#' ## using the `setBackend` function.
+#'
+#' ## Load this data as a `Spectra` object (using a `MsBackendOfflineSql`
+#' ## backend)
+#' library(Spectra)
+#' sps <- Spectra(sqlite_db, source = MsBackendOfflineSql(),
+#'     drv = SQLite())
+#' sps
+#'
+#' ## Define sample annotations for the two data files. Adding one column
+#' ## `"file"` that contains the file name of the data files.
+#' df <- data.frame(sample = c("QC1", "QC2"), file = basename(fls))
+#'
+#' ## Add a spectra variable `"file"` to the `Spectra` object with
+#' ## the raw data files' file names to simplify the linking between
+#' ## samples and spectra performed later.
+#' sps$file <- basename(dataOrigin(sps))
+#'
+#' ## Create a MsExperiment with the spectra and sample data.
+#' mse <- MsExperiment(spectra = sps, sampleData = df)
+#'
+#' ## Establish the link (mapping) between samples and spectra
+#' ## using the column `"file"` in the `sampleData` and the spectra
+#' ## variable `"file"`.
+#' mse <- linkSampleData(mse, with = "sampleData.file = spectra.file")
+#' mse
+#'
+#' ## Write sample data (and the sample to spectra mapping) to the
+#' ## *MsBackendSql* database.
+#' dbWriteSampleData(mse)
+#'
+#' ## List the tables in the database
+#' dbListTables(con)
+#'
+#' ## Sample data was thus stored to the database.
+#' dbGetQuery(con, "select * from sample_data;")
+dbWriteSampleData <- function(x) {
+    if (!inherits(x, "MsExperiment"))
+        stop("'x' is supposed to be an 'MsExperiment'")
+    if (!length(spectra(x))) return()
+    if (inherits(spectra(x)@backend, "MsBackendSql")) {
+        be <- spectra(x)@backend
+        if (inherits(be, "MsBackendOfflineSql")) {
+            flags <- be@flags
```

Comment:
what are the flags ?

Reply:
that's specific for SQLite - it allows to have the connection either read-only or read/write. Other database systems control that type of things using a user management system (e.g. setting permissions for users for specific databases). SQLite does not provide a user management. By default, we're connecting to the databases read-only - but here we want to actually write something to the database, thus we (temporarily) set the connection to read/write. It's a bit of a hack, but with SQLite that's the only way we can do that.